File: examples/zdt1/main.py
"""ZDT1: minimize two conflicting objectives over 30 variables in [0, 1].

Shows NSGA-II, a fitness function that takes a generation at a time, and the hypervolume of the
final non-dominated front.

    python examples/zdt1/main.py
"""

import numpy as np

import genoxide as gx


def zdt1(x):
    """x has a genome per row; the result, a row of objective values per genome."""
    f1 = x[:, 0]
    g = 1 + 9 * x[:, 1:].mean(axis=1)
    return np.column_stack([f1, g * (1 - np.sqrt(f1 / g))])


nsga2 = gx.Nsga2(
    gx.Real((0.0, 1.0), length=30),
    objectives=["minimize", "minimize"],
    population_size=100,
    crossover=gx.SimulatedBinaryCrossover(15),
    mutation=gx.PolynomialMutation(20, rate=1 / 30),
    seed=1,
)
result = nsga2.run(zdt1, batch=True, evaluations=25_000)
front = result.front_objectives[np.argsort(result.front_objectives[:, 0])]

# the hypervolume of the front, with the reference point (1.1, 1.1): the front is sorted by f1, so
# each point adds the rectangle up to the next point's f1
widths = np.diff(np.append(front[:, 0], 1.1))
hypervolume = np.sum(widths * (1.1 - front[:, 1]))
print(f"{len(front)} solutions on the front, hypervolume {hypervolume:.4f} (the whole front: 0.8767)")

File: examples/zdt1/main.rs
//! ZDT1: minimize two conflicting objectives over 30 variables in [0, 1].
//!
//! Shows NSGA-II on genoxide's ZDT1 test problem, and the hypervolume of the final non-dominated
//! front.
//!
//! ```text
//! cargo run --release --example zdt1
//! ```

use genoxide::Objective::Minimize;
use genoxide::multi::indicator::hypervolume;
use genoxide::multi::problems::{TestProblem, Zdt1};
use genoxide::prelude::*;

fn main() -> Result<()> {
    let problem = Zdt1::new(30);
    let nsga2 = Nsga2::builder(problem.real(), [Minimize; 2])
        .population_size(100)
        .crossover(SimulatedBinaryCrossover::new(15.0)?)
        .mutate(PolynomialMutation::per_gene(1.0 / 30.0, 20.0)?)
        .seed(1)
        .build()?;

    let outcome = MultiEngine::new(nsga2, problem)
        .stop_when(Stop::evaluations(25_000))
        .run()?;

    // the hypervolume of the front, with the reference point (1.1, 1.1)
    let front = outcome.front_values();
    let volume = hypervolume(&front, &[1.1, 1.1], &[Minimize; 2]);
    println!(
        "{} solutions on the front, hypervolume {volume:.4} (the whole front: 0.8767)",
        front.len()
    );
    Ok(())
}
